core-tests/shared/src/test/scala/zio/stm/STMSpec.scala

```diff
@@ -1,147 +1,423 @@
+          }
+        ),
+        suite("Make a new `TRef` and")(
+          testM("get its initial value") {
+            (for {
+              intVar <- TRef.make(14)
+              v      <- intVar.get
+            } yield assert(v, equalTo(14))).commit
+          },
+          testM("set a new value") {
+            (for {
+              intVar <- TRef.make(14)
+              _      <- intVar.set(42)
+              v      <- intVar.get
+            } yield assert(v, equalTo(42))).commit
+          }
+        ),
+        suite("Using `STM.atomically` perform concurrent computations")(
+          testM("increment `TRef` 100 times in 100 fibers") {
+            for {
+              tVar  <- TRef.makeCommit(0)
+              fiber <- ZIO.forkAll(List.fill(10)(incrementVarN(99, tVar)))
+              _     <- fiber.join
+              value <- tVar.get.commit
+            } yield assert(value, equalTo(1000))
+          },
+          testM(
+            "compute a `TRef` from 2 variables, increment the first `TRef` and decrement the second `TRef` in different fibers"
+          ) {
+            for {
+              tVars <- STM
+                        .atomically(
+                          TRef.make(10000) <*> TRef.make(0) <*> TRef.make(0)
+                        )
+              tvar1 <*> tvar2 <*> tvar3 = tVars
+              fiber                     <- ZIO.forkAll(List.fill(10)(compute3VarN(99, tvar1, tvar2, tvar3)))
+              _                         <- fiber.join
+              value                     <- tvar3.get.commit
+            } yield assert(value, equalTo(10000))
+          }
+        ),
+        suite("Using `STM.atomically` perform concurrent computations that")(
+          suite("have a simple condition lock should suspend the whole transaction and")(
+            testM("resume directly when the condition is already satisfied") {
+              for {
+                tvar1 <- TRef.makeCommit(10)
+                tvar2 <- TRef.makeCommit("Failed!")
+                join <- (for {
+                         v1 <- tvar1.get
+                         _  <- STM.check(v1 > 0)
+                         _  <- tvar2.set("Succeeded!")
+                         v2 <- tvar2.get
+                       } yield v2).commit
+              } yield assert(join, equalTo("Succeeded!"))
+            },
+            testM(
+              "resume directly when the condition is already satisfied and change again the tvar with non satisfying value, the transaction shouldn't be suspended."
+            ) {
+              for {
+                tvar <- TRef.makeCommit(42)
+                join <- tvar.get.filter(_ == 42).commit
+                _    <- tvar.set(9).commit
+                v    <- tvar.get.commit
+              } yield assert(v, equalTo(9)) && assert(join, equalTo(42))
+            },
+            testM("resume after satisfying the condition") {
+              val barrier = new Barrier
```

Comment:
Barrier definition in util object. Could please tell me what do you think, orginally there was `unsafeRunSync` inside `STM.succeed`